gcloud-java-pubsub/src/main/java/com/google/gcloud/pubsub/spi/PublisherApi.java

```diff
@@ -0,0 +1,498 @@
+    "https://www.googleapis.com/auth/pubsub"
+  };
+
+  public static final PathTemplate PROJECT_PATH_TEMPLATE =
+      PathTemplate.create("/projects/{project}");
+  public static final PathTemplate TOPIC_PATH_TEMPLATE =
+      PathTemplate.create("/projects/{project}/topics/{topic}");
+
+  // ========
+  // Members
+  // ========
+
+  private final ManagedChannel channel;
+  private final ServiceApiSettings settings;
+
+  // ===============
+  // Factory Methods
+  // ===============
+
+  /**
+   * Constructs an instance of PublisherApi with default settings.
+   */
+  public static PublisherApi create() throws IOException {
+    return create(new ServiceApiSettings());
+  }
+
+  /**
+   * Constructs an instance of PublisherApi, using the given settings. The channels are created based
+   * on the settings passed in, or defaults for any settings that are not set.
+   */
+  public static PublisherApi create(ServiceApiSettings settings) throws IOException {
+    return new PublisherApi(settings);
+  }
+
+  private PublisherApi(ServiceApiSettings settings) throws IOException {
+    ServiceApiSettings internalSettings = ApiUtils.settingsWithChannels(settings,
+        SERVICE_ADDRESS, DEFAULT_SERVICE_PORT, ALL_SCOPES);
+    this.settings = internalSettings;
+    this.channel = internalSettings.getChannel();
+  }
+
+  // ==============================
+  // Resource Name Helper Functions
+  // ==============================
+
+  public static final String createProject(String project) {
+    return PROJECT_PATH_TEMPLATE.instantiate(
+        "project", project);
+  }
+
+  public static final String createTopic(String project, String topic) {
+    return TOPIC_PATH_TEMPLATE.instantiate(
+        "project", project,"topic", topic);
+  }
+
+
+  // ========
+  // Getters
+  // ========
+
+  public Channel getChannel() {
+    return channel;
+  }
+
+
+  // =============
+  // Service Calls
+  // =============
+
+  // ----- createTopic -----
+
+  // AUTO-GENERATED DOCUMENTATION AND METHOD - see instructions at the top of the file for editing.
+  /**
+   * Creates the given topic with the given name.
+   *
+   * @param name The name of the topic. It must have the format
+   * `"projects/{project}/topics/{topic}"`. `{topic}` must start with a letter,
+   * and contain only letters (`[A-Za-z]`), numbers (`[0-9]`), dashes (`-`),
+   * underscores (`_`), periods (`.`), tildes (`~`), plus (`+`) or percent
+   * signs (`%`). It must be between 3 and 255 characters in length, and it
+   * must not start with `"goog"`.
+   */
+  public Topic createTopic(String name) {
+    Topic request =
+        Topic.newBuilder()
+        .setName(name)
+        .build();
+
+    return createTopic(request);
+  }
+
+  // AUTO-GENERATED DOCUMENTATION AND METHOD - see instructions at the top of the file for editing.
+  /**
+   * Creates the given topic with the given name.
+   *
+   * @param request The request object containing all of the parameters for the API call.
+   */
```

Comment:
I dug into the gRPC source code to figure this out. It looks like if the error originates from the server, it will be a StatusRuntimeException. Reference:

* https://github.com/grpc/grpc-java/blob/master/stub/src/main/java/io/grpc/stub/ClientCalls.java
* https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/StatusRuntimeException.java

The Status object (which has an error code) can be recovered from a StatusRuntimeException using Status.fromThrowable:

https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/Status.java#L328

It looks like there are other RuntimeExceptions that can be thrown too, particularly for the Iterator, which can throw other RuntimeExceptions if it is mis-used. I imagine there are other scenarios that will throw RuntimeExceptions also.

Why do you need to know which methods are retryable? The ApiCallable can take care of retries.